[lldb] Add SBProcess methods for get/set/use address masks

jasonmolenda opened the pull request:
I'm reviving a patch from phabracator, https://reviews.llvm.org/D155905 which was approved but I wasn't thrilled with all the API I was adding to SBProcess for all of the address mask types / memory regions. In this update, I added enums to control type address mask type (code, data, any) and address space specifiers (low, high, all) with defaulted arguments for the most common case.

This patch is also fixing a bug in the "addressable bits to address mask" calculation I added in AddressableBits::SetProcessMasks. If lldb were told that 64 bits are valid for addressing, this method would overflow the calculation and set an invalid mask.  Added tests to check this specific bug while I was adding these APIs.

rdar://123530562

jasonmolenda committed:
[lldb] Add SBProcess methods for get/set/use address masks

I'm reviving a patch from phabracator, https://reviews.llvm.org/D155905
which was approved but I wasn't thrilled with all the API I was
adding to SBProcess for all of the address mask types / memory
regions. In this update, I added enums to control type address mask
type (code, data, any) and address space specifiers (low, high,
all) with defaulted arguments for the most common case.

This patch is also fixing a bug in the "addressable bits to address
mask" calculation I added in AddressableBits::SetProcessMasks.
If lldb were told that 64 bits are valid for addressing, this method
would overflow the calculation and set an invalid mask.  Added tests
to check this specific bug while I was adding these APIs.

rdar://123530562

DavidSpickett commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,129 @@ class LLDB_API SBProcess {
+  /// Set the number of bits used for addressing in this Process.
+  ///
+  /// In some environments, the number of bits that are used for addressing
+  /// is the natural representation insted of a mask; this method calculates
```
"this method uses that number to calculate the addressing mask that lldb uses internally."

So that "that number" unambiguously refers to the number of addressable bits.

DavidSpickett commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,129 @@ class LLDB_API SBProcess {
+  ///     eAddressMaskTypeAll can be set for both types of addresses.
+  ///     An example where they could be different is a target where
+  ///     instructions are word aligned, so the low 2 bits are always
+  ///     zero.
```
"so the low 2 bits of the PC are always zero"

Otherwise it's a bit of a "so what" statement. Mention the PC so it's more obvious that the code mask would be different in this case.

DavidSpickett commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,129 @@ class LLDB_API SBProcess {
+  ///     zero.
+  ///
+  /// \param[in] num_bits
+  ///     Number of bits that are used for addressing.  e.g. the low 42
```
Maybe phrase this like:
e.g. 42 means that the least significant 42 bits are used for....

The important thing here is that this is not total bits used for addressing, this is a range from bit 0 to some N.

hawkinsw reviewed (commented):
I really appreciate the thorough documentation you wrote for these new functions. Because there is so much overlap in the documentation between the functions, could we refactor it somehow (not sure how?) so that any future change could be more easily tracked?

Just a question. Again, I appreciate your thorough documentation!
Will

jasonmolenda committed:
Doxygen group these methods in SBProcess.h, add longer form definitions
for the `type` and `addr_range` arguments that are used by all of
them, and refer back to those longer descriptions in each individual
message.

Fix the descriptions as suggested by David and Will.

jasonmolenda commented:
Thanks so much for reading through these @DavidSpickett and @hawkinsw !  @adrian-prantl and @JDevlieghere  suggested using a doxygen group for this set of methods and having the long definitions of `type` and `addr_range` a single time, referring back to them from the individual methods, I wasn't thrilled with all that duplicated text either.  I think I did this well enough?

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  /// before reading from memory.
+  ///
+  /// \param[in] type
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  ///     data masks are the same on a given target.
+  ///
+  /// \param[in] addr_range
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  /// before reading from memory.
+  ///
+  /// \param[in] type
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  /// In some environments, the number of bits that are used for addressing
+  /// is the natural representation instead of a mask, but lldb
+  /// internally represents this as a mask.  This method calculates
+  /// the addressing mask that lldb uses that number of addressable bits.
```
I am not 100% sure what you are attempting to convey here (and it may be perfectly correct), but I am having a hard time parsing this sentence.

jasonmolenda replied:
Thanks, let me rewrite it to say "darwin likes addressable bits" directly :) 
On Linux the kernel provides a mask to apply to addresses to remove metadata, it's the most general representation so that's what we use internally in lldb.  But with Darwin on AArch64, we're only concerned with the low order number of bits that are used for addressing -- the addressable bits.  It's possible to calculate the mask incorrectly (I'm fixing my own bug with addressable-bits==64 in this patch), so I want to provide API for environments like Darwin where that's what the data they'll have from the OS etc.

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  ///
+  /// \param[in] num_bits
+  ///     Number of bits that are used for addressing.
+  ///     A value of 42 indicates that the low 42 bits are relevant for
```
```suggestion
  ///     For example, a value of 42 indicates that the low 42 bits are relevant for
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  /// \param[in] num_bits
+  ///     Number of bits that are used for addressing.
+  ///     A value of 42 indicates that the low 42 bits are relevant for
+  ///     addressing, and that higher order bits may be used for various
```
```suggestion
  ///     addressing, and that higher-order bits may be used for various
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  ///     The address that should be cleared of non-addressable bits.
+  ///
+  /// \param[in] type
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  ///
+  /// \a type
+  /// All of the methods in this group take \a type argument
+  /// which is an AddressMaskType enum value.
```
Can our Doxygen system make a reference to the documentation for these named constants? Just a question.

jasonmolenda replied:
I need to read the Doxygen documentation more closely, I just made this up when I was trying to update to having references to these two documentation paragraphs, I don't know if it would render as anything meaningful.

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  /// which is an AddressMaskRange enum value.
+  /// Needing to specify the address range is highly unusual, and the
+  /// default argument can be used in nearly all circumstances.
+  /// On some architectures like AArch64, it is possible to have
```
```suggestion
  /// On some architectures (e.g., AArch64), it is possible to have
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  ///     should be set to 1 in the mask.
+  ///
+  /// \param[in] addr_range
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  /// the addressing mask that lldb uses that number of addressable bits.
+  ///
+  /// \param[in] type
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw commented on lldb/include/lldb/API/SBProcess.h:
```diff
@@ -407,6 +407,117 @@ class LLDB_API SBProcess {
+  ///     metadata like pointer authentication, Type Byte Ignore, etc.
+  ///
+  /// \param[in] addr_range
+  ///     See \ref Mask Address Methods descripton of this argument.
```
```suggestion
  ///     See \ref Mask Address Methods description of this argument.
```

hawkinsw reviewed (commented):
As I said before, I really appreciate you doing such in-depth documentation. I hope these little suggestions help!

jasonmolenda committed:
Fix typeos spotted by Will, update SetAddressableBits doc

jasonmolenda commented:
Thanks for the second round of feedback @hawkinsw .  Let me try to read the Doxygen docs a little more closely tonight and see if the references I threw in there might actually do what I hoped they would.  I briefly looked at the Doxygen docs to see the Grouping feature and got a little overwhelmed and wasn't sure what the expected formatting would be.

JDevlieghere commented on lldb/source/API/SBProcess.cpp:
```diff
@@ -1255,6 +1255,95 @@ lldb::SBFileSpec SBProcess::GetCoreFile() {
+addr_t SBProcess::GetAddressMask(AddressMaskType type,
+                                 AddressMaskRange addr_range) {
+  LLDB_INSTRUMENT_VA(this, type, addr_range);
+  addr_t default_mask = 0;
```
Would it be worth making this a constant (e.g. `LLDB_DEFAULT_ADDRESS_MASK`) so that callers can verify whether they were given a valid address mask?

DavidSpickett replied:
Wouldn't we need a `LLDB_INVALID_ADDRESS_MASK` as well? It can happen that the address mask is valid, and has a value of 0.

jasonmolenda replied:
A mask value of 0 would be an invalid address mask -- no bits are used for addressing.  The variable name "default_mask" was a poor choice.  But explicitly returning LLDB_INVALID_ADDRESS_MASK if we don't find any value might be the correct choice, it would mean you passed an invalid enum argument (or the SBProcess can't access its Process)

JDevlieghere commented on lldb/source/API/SBProcess.cpp:
```diff
@@ -1255,6 +1255,95 @@ lldb::SBFileSpec SBProcess::GetCoreFile() {
 
+addr_t SBProcess::GetAddressMask(AddressMaskType type,
+                                 AddressMaskRange addr_range) {
+  LLDB_INSTRUMENT_VA(this, type, addr_range);
```
`lldb-instr` always puts a newline after `LLDB_INSTRUMENT_VA`. We should do the same here to be consistent with the rest of the SB API.

hawkinsw commented:
> Thanks for the second round of feedback @hawkinsw . Let me try to read the Doxygen docs a little more closely tonight and see if the references I threw in there might actually do what I hoped they would. I briefly looked at the Doxygen docs to see the Grouping feature and got a little overwhelmed and wasn't sure what the expected formatting would be.

Not a problem at all! Just trying to help!!

Will

DavidSpickett commented on lldb/source/Utility/AddressableBits.cpp:
```diff
@@ -33,18 +33,26 @@ void AddressableBits::SetHighmemAddressableBits(
 }
 
+addr_t AddressableBits::AddressableBitToMask(uint32_t addressable_bits) {
+  assert(addressable_bits <= sizeof(addr_t) * 8);
```
If we're in release mode, this assert isn't here. Is it possible for the user to set addresable_bits to > 64 via `settings set`?

If they do it's a bit of a garbage in garbage out situation, but wondering if we could add a check at that level too, or already have.

jasonmolenda replied:
I forgot that I was duplicating the addressable-bits-to-mask calculation in Process.cpp when I read the settings values.  I pushed a change to use the new static method in AddressableBits which includes the assert.

jasonmolenda committed:
Apply suggestions from code review

Add Will's good copyedit suggestions.

Co-authored-by: Will Hawkins <whh8b@obs.cr>

jasonmolenda committed:
Address Jonas' suggestions, add LLDB_INVALID_ADDRESS_MASK

Have SBProcess::GetAddressMask return LLDB_INVALID_ADDRESS_MASK if
no mask can be retrieved.

jasonmolenda committed:
Incorporate David's documentation improvements

jasonmolenda committed:
Use static method to compute mask from bits for settings

David asked if I should assert on >64 addressable bits from
the user settings, and I saw that I failed to use the new
static method in AddressableBits when creating a mask from
the settings, which has the assert it in.

Also make the API test a little easier to read for the
eAddressMaskTypeCode.

jasonmolenda committed:
Fix LLDB_INVALID_ADDRESS_MASK value

I had this set to 0, which means "all bits are used for an address",
i.e. no masking will be perfomed, which is an acceptable state not
an invalid one.

JDevlieghere reviewed (approved):
LGTM but please fix the formatting issue before merging.

jasonmolenda committed:
reflow comments

jasonmolenda commented:
Temporarily reverted this change while I investigate why the tests failed on all the linux bots (lldb-x86_64-debian, lldb-arm-ubuntu, lldb-aarch64-ubuntu), I'll build up in a VM and debug.

jasonmolenda commented:
It does occur to me that I'm going to need to only run this API test on targets which have a FixAddress method in their ABI, the base class own't do it.  Maybe it should have a base class impl that can be overridden, and use the Process masks if they are set.  (they're all initialized to 0 which is an impossible mask value (no address bits))  The base class impl would need to be overridden for different architectures, e.g. on AArch64 where TBI or MTE are used, in addition to clearing/setting the top byte, we need to use b55 to determine if the non-address bits are set to 1 or 0.  On armv7 the 0th bit can be used for metadata on TypeCode fixes, etc.

But still I'd expect the lldb-aarch64-ubuntu bot to have had a FixAddress impl in its ABI.  will check out when I build it up.

jasonmolenda commented:
I think I'm going open a new PR with the base class address masking added to the patch.  I think having these API and the unwritten caveat is "they may be no-ops if you're using an ABI that doesn't do FixAddress" is going to confuse people.  I still want to investigate why my API test failed on aarch64 linux because I didn't expect that, it's possible there's something else going on.  But the ABI thing is definitely something I think should be addressed.

jasonmolenda commented:
After debugging, updating, and testing on aarch64-unbuntu, x86_64-macos, and arm64-macos, I have created a new PR with this commit plus an additional commit to fix the issues I found on the different platforms. https://github.com/llvm/llvm-project/pull/83663
